### ISSUE TITLE
Calling F# from javascript [Newbie question]

### ISSUE THREAD

**Issue by whitetigle** [fsprojects/Fable#234]
Hi,
I am relatively new to f# and I am not a javascript specialist (meaning: I do use javascript and have been doing a lot of inetrfacing from other languages such as scala or haxe but don't write code using require.js like what is generated by fable, etc...). 

Playing with the samples, I understand how to create a whole app in f# and launch it trough the use of require. 

Now my problem is that I would also like to have let's say some nice f# functions I could call straight from javascript. How would I do that?

The documentation says :
> The F# compiler compiles type and module members as usual for an OOP language like C#, no matter whether the arguments are curried or tupled. Thanks to this, they can be called "normally" from an external language like C# or JavaScript. For example:

However, how do I make this "normal" call work ?

If I understand the sample we would have a file containing the following definitions:

> module MyModule =
>     let myCurriedFunction a b = a + b
>     let myTupledFunction(a, b) = a + b
> 
> type MyType() =
>     member x.MyCurriedMethod a b = a + b
>     member x.MyTupledMethod(a, b) = a + b

Then it would just be a matter of (for instance?)
`
```
<script>
var lib = require("lib/myFSLib");
console.log(lib.MyModule.myCurriedFunction(2, 2));
</script>
```
`

Given this example, 

1. Where would do I put the code of the lib? in a .fs file? .fsx?
2. What would be the name of this file?
3. Where would this file saved?
4. How would I compile this file to javascript? (fable?)
5. Then, how would I call it from javascript? (require?)

I apologise in advance since I think there's quite a few concepts I don't yet grasp.
I have enjoyed the use of f# so far and would really like to use it in my apps like I do with scala.
Thanks in advance for your replies!

**Comment by alfonsogarciacaro** [comment]
Hi there and thanks for giving Fable a try! As Fable integrates with modern JS workflows, it's difficult to explain how it works without a solid understanding of JS module systems: es2015 (aka es6), commonjs, amd, umd... I recommend you to read [this post](https://www.airpair.com/javascript/posts/the-mind-boggling-universe-of-javascript-modules) for that ([this one](https://addyosmani.com/writing-modular-js/) is also useful but a bit more advanced).

Most of your questions are answered in the [docs](http://fsprojects.github.io/Fable/docs.html) but I'll try to give a summary here:

1. You can use either an `.fs` or `.fsx` file, both will generate the same `.js` file. The main advantage of `.fsx` is you don't need a `.fsproj` file.
2. Any name you want :wink: The name will be the same in JS, only the extension will change.
3. In the output directory you specify (same as the F# project directory if you don't use the `--outDir` compiler argument). Note the folder structure within the project directory will be kept. For example, if you have an F# project that looks like this (assuming `main.fsx` includes a `#load "util/util.fs"` directive):

```
src/util/util.fs
src/main.fsx
```

And you compile it using `fable src/main.fsx --outDir out`, Fable will generate the following files:

```
out/util/util.js
out/main.js
```

4. Of course you'll use Fable to compile F# to JS! Please read [Compiling to JavaScript](http://fsprojects.github.io/Fable/docs/compiling.html) for more info. If you want to call the generated code from JS it's important you specify the module system you want to use with the [`--module` argument](http://fsprojects.github.io/Fable/docs/compiling.html#Modules). By the default, universal modules (`umd`) are used, which are compatible with both `amd` and `commonjs`.

5. As explained above, it depends on the module system you're using. When writing a web app, if you're not using a module bundler like Webpack or Browserify, normally you would load AMD modules with [require.js](http://requirejs.org/docs/start.html). Not to be confused with the `require` command of `commonjs`... it's complicated I know, welcome to modern JS development 😉 

Note that if your F# file only contains one module, for your convenience Fable will flatten it so you don't need to qualify all member calls (that is, in your example you wouldn't need to write `MyModule`, it'll be enough with `lib.MyModule.myCurriedFunction(2, 2)`). More info [here](http://fsprojects.github.io/Fable/docs/compiling.html#Modules).

In conclusion, if using `require.js` from a `<script>` tag, your example above would become:

```html
<script data-main="scripts/main.js" src="scripts/require.js"></script>
<script>
requirejs(["./lib/myFSLib"], function (lib) {
    console.log(lib.myCurriedFunction(2, 2));
}
</script>
```

> Note that we put `./` in front of the path to indicate it's relative and we omit the `.js` extension.

I hope this helps. Please give it a try and when stuck, try to review the docs for info and if you cannot find the answer there, post the question here :+1:

Have fun!

**Comment by whitetigle** [comment]
Excellent! 
I just updated my files and I confirm that I can call my f# lib from javascript. 
All your instructions were very clear. And I have learnt yet another lot of stuff today ;)
Bravo and thanks!

**Comment by whitetigle** [comment]
Hi again!
So yesterday I solved my problem using the AMD/requirejs approach. 

Now I have compiled my code using UMD. Furthermore I have bundled my js code using browserify and loaded the generated code using

`  <script src="./out/bundle.js"></script>
`
My code is correctly called when the page loads: I have a printfn inside that works.

But, I have a function inside that I would like to call from my web page
`let doSomething = printfn "doSomething"`

The generated code gives:
```
  var doSomething = exports.doSomething = _fableCore.String.fsFormat("doSomething")(function (x) {
    console.log(x);
  });

```
Since I do not have lib context anymore, my question may sound silly, but where is my doSomething function exported? Which context? In the window object? Do I have to make some manual binding?

Thanks again for your patience and your help!

**Comment by alfonsogarciacaro** [comment]
As code generated by Fable is exported using JS modules, it must be imported using modules too so the context is defined by the way you import it (it's not enough to include it in a `<script>` tag in your HTML). For example:

```js
// amd
requirejs(["./out/bundle.js"], function (myLib) {
   myLib.doSomething();
}

// commonjs
var myLib = require("./out/bundle.js");
myLib.doSomething();

// es2015
import * as myLib from "./out/bundle.js";
myLib.doSomething();
```

But you need to change your function in F# as right now `doSomething` is just an empty value (`unit` the result of `printfn` being called eagerly):

```fsharp
let doSomething() = printfn "doSomething"
```

**Comment by whitetigle** [comment]
Following the links you gave me yersterday I have then been reading a lot of stuff regarding modules but struggled to connect all the dots. Now all seems pretty clear. 

I will definitely make my mind around the modular approach and level up my js skills. There are quite a few concepts I need to grasp in the modern js approach.

Thanks again for your kind help and your patience.
